### PR TITLE
[Ide] Adds feature in MonoDevelop to pin opened tabs

### PR DESCRIPTION
**Pull request by netonjm**
## Feature

* Enables our document bar to **pin tabs**.
* Adds some menu items in the contextual menu like: "Close All Except Pinned", "Un/Pin Tab"
* Parity with VS windows

| |  | |

dragdrop

This feature allows to lock some of your most used windows (tabs)
to easily detect and open like VS2013 added.

Fixes VSTS #890509 - Add support for pinned tabs
https://devdiv.visualstudio.com/DevDiv/_workitems/edit/890509

Fixes VSTS #945511 - Use the same pin from Start window at a smaller size for Tab pinning experience

## How to test

1) Feature is enabled by default
2) When enabled a pin icon will appear in every tab on over
3) Try pin/unpin tabs and ensure behaviour is correct
the reordering only works over same kind of tabs groups
4) Try the new commands created:
- Close All Except pinned
- Pin Tab
5) Test commands modified
- Close All
6) Ensure commands don't work with feature disabled

Fixes VSTS #890509 - Add support for pinned tabs
https://devdiv.visualstudio.com/DevDiv/_workitems/edit/890509
Fixes VSTS #890508 - Add support for pinned tabs
https://devdiv.visualstudio.com/DevDiv/_workitems/edit/890508
Includes fixes from:
#2487 and #2449 and #2606

## Mac
image

## Windows
image

### CODE REVIEW

**Comment by netonjm** [comment]
Rebased on the top of master 👍

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Components.DockNotebook/DockNotebook.cs [hunk 1]
```diff
@@ -111,6 +112,21 @@ public DockNotebook ()
...
+			IdeApp.Preferences.EnablePinnedTabs.Changed += EnablePinnedTabs_Changed;
+		}
+
+		void EnablePinnedTabs_Changed (object sender, EventArgs e) => RefreshCurrentTabStrip ();
+
+		void RefreshCurrentTabStrip ()
+		{
+			tabStrip.IsPinEnabled = IdeApp.Preferences.EnablePinnedTabs.Value;
+			if (!tabStrip.IsPinEnabled) {
+				pages.ToList ().ForEach (d => d.IsPinned = false);
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Components.DockNotebook/DockNotebook.cs]: please use foreach here.

Comment (netonjm) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Components.DockNotebook/DockNotebook.cs]: Fixed

**Review by Therzok — APPROVED** [review]

**Comment by sevoku** [comment]
LGTM!

But do we really need a setting for this? cc @hbons for wording

Also, let's enable it by default? Or is it off by default on VS on Windows?

**Comment by sevoku** [comment]
Per discussion with @hbons on Slack, the new preferences panel is not necessary. On Windows it's common to have a checkbox for each and every little thing, but on other platforms (especially Mac) less is more. We don't see any benefit for the user to be able to disable this feature (same applies to preview tabs) and if we introduce this, we should also add properties for many more Shell features (like split view, floating pads, etc.), such fine graned feature toggles don't make much sense.

If the idea behind this panel was to make it possible to disable it in case it doesn't work as expected (like a feature flag for a preview feature), then we should add a special preview property panel, for such kind of things. But I'd rather do some thorough testing and ship it as it is (without the panel).

**Comment by netonjm** [comment]
Removed the settings panel, enabled by default and rebased 👍  @sevoku @slluis

**Comment by netonjm** [comment]
Green! 💃

**Comment by netonjm** [comment]
rebased

**Comment by Therzok** [comment]
@slluis can we merge this?

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Commands/FileTabCommands.cs [hunk 2]
```diff
@@ -102,34 +118,76 @@ protected override void Run ()
...
-			if (activeDoc == null) {
-				info.Enabled = false;
+		protected override ImmutableArray<Document> GetDocumentExceptions ()
+		{
+			var active = IdeApp.Workbench.ActiveDocument;
+			if (active == null)
+				return ImmutableArray<Document>.Empty;
+			var activeNotebook = ((SdiWorkspaceWindow)active.Window).TabControl;
+
+			var contents = IdeApp.Workbench.Documents.Where (doc => ((SdiWorkspaceWindow)doc.Window).TabControl == activeNotebook && GetTabFromDocument (doc).IsPinned)
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Commands/FileTabCommands.cs]: Nit:
There's a SelectAsArray extension method [here](http://source.roslyn.io/#Microsoft.CodeAnalysis/Collections/ImmutableArrayExtensions.cs,365be1a01db3ca7a)

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Commands/FileTabCommands.cs [hunk 3]
```diff
@@ -102,34 +118,76 @@ protected override void Run ()
...
+	{
+		protected override ImmutableArray<Document> GetDocumentExceptions ()
+		{
+			var active = IdeApp.Workbench.ActiveDocument;
+			if (active == null) {
+				return ImmutableArray<Document>.Empty;
 			}
-
-			info.Enabled = false;
+			return new ImmutableArray<Document> () { active.Window.Document };
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Commands/FileTabCommands.cs]: `ImmutableArray.Create (active.Window.Document);`

**Review by Therzok — APPROVED** [review]

**Comment by netonjm** [comment]
rebased

**Comment by netonjm** [comment]
Failing test is not related

**Comment by jsuarezruiz** [comment]
I have been doing some tests of this branch. I love the idea!.
 My feedback:
- The icon to pin a tab looks nice on both themes (light and dark).
- Reordering tests are satisfactory. You can reorder pinned tabs between them. You can not mix pinned tabs with not pinned ones.
- The biggest problem I have found, is when reopen a project with pinned tabs. Initially they do not remember the state and appear as not pinned. Trying to pin another tab is when it refreshes the state.

fixed-tabs-error

**Comment by netonjm** [comment]
Thanks for the review Javier! Related to store the current pinned tabs yes, I didn't include it because we need to think about where to store that information, I understand that it is related to each project so it may make sense in the userconfiguration, but it needs to be defined. we could open a issue to file it.

**Comment by jmatthiesen** [comment]
Adding @vancura to comment on the icon usage. Vaclav - would love your thoughts on the icon used for a pin that can be tabbed, versus an already pinned tab.

Untabbed pin, ready to tab:
image

Tabbed pin:
image

**Comment by netonjm** [comment]
> Adding @vancura to comment on the icon usage. Vaclav - would love your thoughts on the icon used for a pin that can be tabbed, versus an already pinned tab.
> 
> Untabbed pin, ready to tab:
> image
> 
> Tabbed pin:
> image

Only for the records https://github.com/mono/monodevelop/pull/2449#issuecomment-303248814

**Review by sevoku — CHANGES_REQUESTED** [review]
This works really nice! But see my comments, ready to merge once addressed :)

**Comment by vancura** [comment]
Re icons: do you have the final icons? They're at https://devdiv-design.visualstudio.com/D3%20Studio/_workitems/edit/7382/ (the newer file, Fix.zip).

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Commands/FileTabCommands.cs (hunk 2 above)
Comment (netonjm) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Commands/FileTabCommands.cs]: Fixed

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Components.DockNotebook/DockNotebookTab.cs [hunk 4]
```diff
@@ -113,6 +118,20 @@ class DockNotebookTab: IAnimatable, IDisposable
...
+		bool isPinned;
+		public bool IsPinned { 
+			get { return isPinned; }
+			set {
+				if (isPinned == value)
+				   return;
+				if (OnChangingPinned != null)
+					OnChangingPinned (this, value);
+				isPinned = value;
+				if (OnChangedPinned != null)
```
[the remark this answers is not in the document]
Comment (netonjm) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Components.DockNotebook/DockNotebookTab.cs]: Fixed

File: main/src/core/MonoDevelop.Ide/ExtensionModel/Commands.addin.xml [hunk 5]
```diff
@@ -543,6 +543,18 @@
 			defaultHandler = "MonoDevelop.Ide.Commands.ReopenClosedTabHandler"
 			_label = "Reopen Closed Tab"
 			_description = "Opens the last tab that has been closed"/>
+	<Command id = "MonoDevelop.Ide.Commands.FileTabCommands.CloseAllExceptPinned"
+			defaultHandler = "MonoDevelop.Ide.Commands.CloseAllExceptPinnedHandler"
+			_label = "Close All Except _Pinned"
+			_description = "Close all files except pinned"
+			shortcut = "Control|Shift|P"
```
[the remark this answers is not in the document]
Comment (netonjm) [reply main/src/core/MonoDevelop.Ide/ExtensionModel/Commands.addin.xml]: Fixed

File: main/src/core/MonoDevelop.Ide/ExtensionModel/Commands.addin.xml [hunk 6]
```diff
@@ -543,6 +543,18 @@
...
+			macShortcut = "Meta|Shift|P" />
+	<Command id = "MonoDevelop.Ide.Commands.FileTabCommands.PinTab"
+			defaultHandler = "MonoDevelop.Ide.Commands.PinTabHandler"
+			_label = "_Pin Tab"
+			_description = "Pin/Unpin current Tab selected"
+			shortcut = "Control|Alt|P"
```
[the remark this answers is not in the document]
Comment (netonjm) [reply main/src/core/MonoDevelop.Ide/ExtensionModel/Commands.addin.xml]: Fixed

**Review by sevoku — APPROVED** [review]

**Comment by sevoku** [comment]
@netonjm are the new icons already included?

**Comment by sevoku** [comment]
Oh and what about:

> When reopen a project with pinned tabs do not remember the state and appear as not pinned.

Maybe we could simply store the state in UserPrefs.xml?

**Comment by netonjm** [comment]
Added persistency and new icons 

lastdocumentopened

**Review by sevoku — APPROVED** [review]

**Comment by netonjm** [comment]
👏🏻👏🏻👏🏻👏🏻🍾

**Comment by Therzok** [comment]
Nicely done!

**Pull request merged**